### PR TITLE
Adding intersection cut

### PR DESCRIPTION
**Pull request by SaTahernejad**

### CODE REVIEW

File: include/MibSBilevel.h [hunk 1]
```diff
@@ -47,6 +47,7 @@ class MibSBilevel {
    double *lowerSolutionOrd_;
    double *optLowerSolution_;
    double *optLowerSolutionOrd_;
+   double *lSolution_;
```
Comment (tkralphs) [inline include/MibSBilevel.h]: It looks like `lSolution_` and `optLowerSolution_` are storing the same thing here. Any reason you added `lsolution_`? In any case, try to use similar naming conventions through (spelling out `lower` versus just prepending an `l` to indicate that some data element is associated with the lower level problem).

File: include/MibSCutGenerator.h [hunk 2]
```diff
@@ -70,6 +70,17 @@ class MibSCutGenerator : public BlisConGenerator {
    /** Add Benders-type cuts for zero sum problems **/
    int bendersZeroSumCuts(BcpsConstraintPool &conPool);
 
+   /** Add intersection cuts for discrete problems **/
+   int intersectionCuts(BcpsConstraintPool &conPool, double *lowerSolution);
+
+   void getAlpha(double** extRay, double* lowerSolution, int numStruct,
```
Comment (tkralphs) [inline include/MibSCutGenerator.h]: Maybe give this a name to make it clear that it is a helper function for intersection cuts (maybe `getAlphaIntersectionCut`). What we really should do is to create a separate C++ class for each class of cuts. Perhaps we could even use CGL directly.

File: include/MibSCutGenerator.h [hunk 3]
```diff
@@ -70,6 +70,17 @@ class MibSCutGenerator : public BlisConGenerator {
...
+		 int numNonBasic, const double* lpSol,
+		 std::vector<double> &alphaVec);
+
+   double solveModel(const CoinPackedMatrix* matrix, double** extRay,
```
Comment (tkralphs) [inline include/MibSCutGenerator.h]: Here again (and especially in this particular case), the name should indicate that this is associated with intersection cut.

File: include/MibSBilevel.h (hunk 1 above)
Comment (SaTahernejad) [inline include/MibSBilevel.h]: Yes, you are right. As I checked defining ` lSolution_` is useless. I will modify it.

Sure, I will change all names with` l` prefix.

File: include/MibSCutGenerator.h (hunk 2 above)
Comment (SaTahernejad) [inline include/MibSCutGenerator.h]: I will modify its name. 
Yes, it seems good to have separate class for each cuts. Do you mean that I work on it now or change it whenever we start rewriting MibS again?

File: include/MibSCutGenerator.h (hunk 3 above)
Comment (SaTahernejad) [inline include/MibSCutGenerator.h]: Ok, I will change it.

File: include/MibSCutGenerator.h (hunk 2 above)
Comment (tkralphs) [inline include/MibSCutGenerator.h]: Just make a note of it on the global TODO list and we'll think about it later. We need to move MibS over to DisCO at some point and then we can think about re-design.

File: include/MibSCutGenerator.h [hunk 4]
```diff
@@ -71,15 +71,17 @@ class MibSCutGenerator : public BlisConGenerator {
    int bendersZeroSumCuts(BcpsConstraintPool &conPool);
 
    /** Add intersection cuts for discrete problems **/
-   int intersectionCuts(BcpsConstraintPool &conPool, double *lowerSolution);
+   int intersectionCuts(BcpsConstraintPool &conPool, double *optLOowerSolution);
```
Comment (tkralphs) [inline include/MibSCutGenerator.h]: You have a typo here that is repeated in several other places.

File: src/MibSCutGenerator.cpp [hunk 5]
```diff
@@ -562,74 +560,74 @@ MibSCutGenerator::intersectionCuts(BcpsConstraintPool &conPool, double *lowerSol
...
+MibSCutGenerator::getAlphaIntersectionCut(double** extRay, double* optLOowerSolution,
+					  int numStruct, int numNonBasic,
+					  const double* lpSol, std::vector<double> &alphaVec)
 {
     
     OsiSolverInterface * oSolver = localModel_->solver();
     
     const CoinPackedMatrix * matrix = oSolver->getMatrixByRow();
-    double * lObjCoeffs = localModel_->getLowerObjCoeffs();
+    double * lowerObjCoeffs = localModel_->getLowerObjCoeffs();
```
Comment (tkralphs) [inline src/MibSCutGenerator.cpp]: I didn't really mean that you had to change `l` to `lower` in every single instance, just that in that particular case, there were two variables declared for similar purposes right next to each other that used different conventions, so it was particularly egregious.

File: include/MibSCutGenerator.h (hunk 4 above)
Comment (SaTahernejad) [inline include/MibSCutGenerator.h]: Oh, you are right. I will modify that.

File: src/MibSCutGenerator.cpp (hunk 5 above)
Comment (SaTahernejad) [inline src/MibSCutGenerator.cpp]: I will change them again by using `l ` to be similar to the other parts of MibS.

**Comment by tkralphs** [comment]
This has been merged (see 1c83ba4).

**Pull request closed without merging**